Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/ApromoreZKListener.java

```diff
@@ -0,0 +1,200 @@
+            LOGGER.error("Exception when verifying digital signature of JWT", e);
+
+            return false;
+        }
+    }
+
+    /**
+     * Issue a new JWT with an extended expiry.
+     */
+    private static void refreshSessionTimeout(
+            final JWTClaimsSet preExistingJwtClaimsSet,
+            final Execution exec,
+            final String existingJwtStr,
+            final HttpServletResponse httpServletResponse) {
+        LOGGER.debug(">>>>> Refreshing session timeout");
+        try {
+            final ConfigBean config = (ConfigBean) SpringUtil.getBean("portalConfig");
+            final String securityMsHost = config.getSecurityMsHost();
+            final String securityMsPort = config.getSecurityMsPort();
+            LOGGER.debug("Using securityMsHost {} and securityMsPort {}", securityMsHost, securityMsPort);
+            final String refreshTokenUrl = "http://" + securityMsHost + ":" + securityMsPort + "/refreshJwtToken";
+            LOGGER.debug("Using refreshTokenUrl {}", refreshTokenUrl);
+
+            final RefreshTokenResponse refreshTokenResponse = ClientBuilder.newClient()
+                .target(refreshTokenUrl)
+                .path(existingJwtStr)
+                .request(MediaType.APPLICATION_JSON)
+                .get(RefreshTokenResponse.class);
+
+            LOGGER.debug("Refreshed session timeout, refreshTokenResponse {}", refreshTokenResponse);
+
+            final String updatedAuthHeader = refreshTokenResponse.getAuthHeader();
+            final String updatedSignedAuthHeader = refreshTokenResponse.getSignedAuthHeader();
+            LOGGER.info(">>> updatedAuthHeader {}", updatedAuthHeader);
+
+            final JWT refreshedJwt = JWTParser.parse(updatedAuthHeader);
+            final JWTClaimsSet refreshedJwtClaimsSet = refreshedJwt.getJWTClaimsSet();
+            LOGGER.debug("refreshedJwtClaimsSet {}", refreshedJwtClaimsSet);
+            LOGGER.debug("refreshedJwtClaimsSet.getExpirationTime() {}", refreshedJwtClaimsSet.getExpirationTime());
+
+            boolean expiresSoon = JwtHelper.doesJwtExpiryWithinNMinutes(
+                    preExistingJwtClaimsSet, 5);
+            LOGGER.debug(">>> expiresSoon {}", expiresSoon);
+
+            JwtHelper.writeCookie(httpServletResponse, JwtHelper.AUTH_HEADER_KEY, updatedAuthHeader);
+            JwtHelper.writeCookie(httpServletResponse, JwtHelper.SIGNED_AUTH_HEADER_KEY,
+                    updatedSignedAuthHeader);
+            LOGGER.debug("Updated cookie values written to the HttpServletResponse");
+        } catch (final Exception e) {
+            LOGGER.warn("Unable to refresh session timeout", e);
+        }
+    }
+
+    /**
+     * Broadcast that this session has been signed out.
+     */
+    private static void signOut(final Session session) {
+        EventQueues.lookup("signOutQueue", EventQueues.APPLICATION, true)
+                .publish(new Event("onSignout", null, session));
+    }
+}
```

Comment:
It looks like the onSignout event has been broken for some time; it no longer closes the user's other tabs when the user signs out.  (This feature depended on unstandardized Javascript behavior, so it might just be that browsers are more strict nowadays about letting scripts close windows.)